Import numeric initial data in GeneralizedHarmonic executable

nilsvu opened the pull request:
## Proposed changes

Adds support for importing numeric initial data to the `evolution::DgElementArray` and the GeneralizedHarmonic executable. Other executables can follow this example to add numeric initial data as well.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Support importing numeric initial data in DgElementArray

nilsvu committed:
Support non-analytic initial data in GH executable

nilsvu committed:
Use boundary_condition_tag in ImposeBoundaryConditions

Remove uses of initial_data_tag from the file. Executables may not
have such a tag, e.g. if they load numeric initial data from a file.

geoffrey4444 commented:
Happy to review! Is it rebase on master after the main numerical import code got merged?

nilsvu commented:
yep this is rebased, and you can rebase your private branches on this as well if you like. I'm labeling as _in progress_ just while making sure everything works.

nilsdeppe commented:
Can we close this? It seems to be a duplicate of #1714 which is replaced by #2005 It would seem better to coordinate rather than opening competing PRs

geoffrey4444 commented:
To be fair, we did discuss this. @nilsleiffisher didn’t go rogue! :) It’s just that I thought he had something like #2005 when we discussed, but actually #1986 is more complicated. I haven’t tried it out yet but have used #2005 (and #1714) extensively. Maybe we need to see which ideas from #1986 belong in #2005?

Partly, this came up again because I used #2005 as a basis for trying CCE with GH. I used this one because it changed fewer things and because I’ve done lots of runs with it.

nilsvu committed:
Support importing numeric initial data in GH executable

nilsvu committed:
Add NumericInitialData protocol

This protocol defines the compile-time interface for classes
that provide numeric initial data for an evolution.

nilsvu committed:
Extract the fields to import numeric initial data for from the evolution system

geoffrey4444 commented on tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_ImposeBoundaryConditions.cpp:
```diff
@@ -38,7 +39,7 @@ struct PrimitiveVar : db::SimpleTag {
 };
 
-struct BoundaryCondition {
+struct BoundaryCondition : MarkAsAnalyticSolution {
```
Am I right that you only have to `MarkAsAnalyticSolution` if you're setting a Dirichlet condition by hand and not from an existing analytic solution, such as KerrSchild? A comment to that effect here might be helpful, if so. Otherwise, could you please document somewhere in this PR when you have to add `MarkAsAnalyticSolution`?

nilsvu replied:
Currently we only support dirichlet boundary conditions imposed by an analytic solution, and since that's being checked in `ImposeBoundaryConditions.hpp` I made this change in the test. You don't need to change anything in `src`. I added a comment

geoffrey4444 commented on src/Evolution/Protocols.hpp:
```diff
@@ -0,0 +1,33 @@
+ * - `import_fields`: The list of tags that should be imported from a volume
+ * data file
+ *
+ * Here's an example for a class that conforms to this protocol:
```
example for -> example of

geoffrey4444 commented on tests/Unit/Evolution/Test_Protocols.cpp:
```diff
@@ -0,0 +1,33 @@
+
+static_assert(
+    evolution::protocols::NumericInitialData<ValidNumericInitialData>::value,
+    "Failed testing protocol");
```
Please make this error message more helpful

nilsvu replied:
this a standard error message for an assertion in `tests/`, see e.g. any file in `tests/Unit/Utilities/TypeTraits/`. Not sure why we have those message at all actually, we could just remove them and have the `static_assert` without the error message.

FYI, I'm changing how protocols work in #2221

geoffrey4444 commented on tests/Unit/Evolution/Test_Protocols.cpp:
```diff
@@ -0,0 +1,33 @@
+    "Failed testing protocol");
+static_assert(not evolution::protocols::NumericInitialData<
+                  InvalidNumericInitialData>::value,
+              "Failed testing protocol");
```
Please make this error message more helpful

geoffrey4444 commented on src/Evolution/Protocols.hpp:
```diff
@@ -0,0 +1,33 @@
+
+/*!
+ * \ingroup ProtocolsGroup
+ * \brief Indicates the `ConformingType` provides compile-time information for
```
Design question: why do you need a protocol for this? What's a use case where you would have more than one struct conforming to this?

nilsvu replied:
My intention is to not couple the numeric data importing code to the system, but allow specifying the fields to be imported by any class that exposes the `import_fields`. So instead of using the `evolution::NumericInitialData<System>` class you could provide a class in the Metavariables that, for example, adds a bunch of diagnostic fields that should be imported from the volume data file to the list of system fields. If the DgElementArray will only ever need the list of fields to import then I could just pass that, but I'm thinking we may want to adjust the input file group or the callback action (e.g. for data verification) in the Metavariables. If we want to do that we can change the protocol and the classes that conform to it instead of changing template parameters in a bunch of places. Does that make sense?

geoffrey4444 commented on tests/Unit/Evolution/Test_NumericInitialData.cpp:
```diff
@@ -0,0 +1,34 @@
+static_assert(
+    test_protocol_conformance<evolution::NumericInitialData<System>,
+                              evolution::protocols::NumericInitialData>,
+    "Failed testing protocol conformance");
```
Please make this error message more specific

geoffrey4444 commented on tests/Unit/Evolution/Test_NumericInitialData.cpp:
```diff
@@ -0,0 +1,34 @@
+// Test that the `import_fields` extracted from the system are correct
+static_assert(std::is_same_v<
+                  typename evolution::NumericInitialData<System>::import_fields,
+                  tmpl::list<FieldTag>>,
```
With more than one tag, does order matter? Maybe test a second system with three tags as well?

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -9,7 +9,6 @@ set(LIBS_TO_LINK
   DomainCreators
   Evolution
   GeneralRelativity
-  GeneralRelativitySolutions
```
Why did you remove this? 

I'm guessing the idea is that if you import numerical initial data, you don't need analytic solutions, but that's not yet true. Because we don't yet have GH constraint preserving boundary conditions, we still set analytic boundary conditions even when reading numerical initial data. As long as the outer boundary is far enough away, this is ok.

Anyway, I'm not opposed to this change in principle, but actually evolving numeric initial data for now requires that the executable still sets an analytic solution for the boundary condition. So maybe this is just putting a framework in place for later on.

nilsvu replied:
I'm just moving where this library is linked from this general place to further down where executable are added that actually use analytic solutions. The reasoning is that not all executables will need to link with `GeneralRelativitySolutions`. But the current executables still do.

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -22,10 +21,24 @@ set(LIBS_TO_LINK
-  EvolutionMetavars
-  "${LIBS_TO_LINK}"
+function(add_generalized_harmonic_executable
+  EXECUTABLE_NAME INITIAL_GUESS BOUNDARY_CONDITIONS)
```
Since this is for evolution, not initial data, shouldn't INITIAL_GUESS -> INITIAL_DATA here?

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -22,10 +21,24 @@ set(LIBS_TO_LINK
+)
+target_link_libraries(
+  EvolveKerrSchild
+  PRIVATE
```
What does PRIVATE do here? I think I have the gist, that for this one executable, the target gets linked against GeneralRelativitySolutions...I'm just unclear what PRIVATE does.

nilsvu replied:
Since this is not a library I don't think it matters whether this is `PRIVATE` or `PUBLIC`. The call to `target_link_libraries` needs a keyword though, so I chose `PRIVATE`. In other words: Since `PUBLIC` is just `PRIVATE + INTERFACE` but `INTERFACE` has no meaning for an executable I think `PRIVATE` is technically the correct choice.

For libraries remember this: link libs with `PRIVATE` when they are used only in cpp files, link them with `INTERFACE` if they are used only in hpp files and link them with `PUBLIC` if they are used in both.

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicFwd.hpp:
```diff
@@ -5,4 +5,19 @@
 
 #include <cstddef>
 
+/// \cond
```
So I can understand, why are these forward declarations needed now but not before this PR?

nilsvu replied:
Because the `EvolutionMetavars` didn't have template parameters before but hard-coded the initial data and boundary conditions. Now they are chosen in cmake, so they need to be forward-declared.

geoffrey4444 commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -14,21 +14,111 @@
+ *
+ * \see `DgElementArray`
+ */
+struct ImportNoInitialData {};
```
This ImportNoInitialData struct seems odd, or at least oddly named. Not sure what a good alternative is. Maybe a better comment explaining when you use this struct.

nilsvu replied:
What do you think is unclear? I guess since I wrote the code I'm not sure what confused someone who reads it for the first time. These structs are only used as template parameters to `DgElementArray` to choose whether or not to import numeric initial data, and to configure the numeric initial data to import. `ImportNoInitialData` is the default.

geoffrey4444 commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -14,21 +14,111 @@
+    ::Actions::SetData<ImportFields>, DgElementArray>;
+
+template <typename DgElementArray, typename ImportInitialData>
+struct import_numeric_data_cache_tags {
```
what does this struct do? please add or point to the relevant documentation. Is this one just used if not importing numeric initial data? If so, should the second template parameter be `typename ImportNoInitialData`?

nilsvu replied:
It's in a detail namespace, so it's an implementation detail. The class template is specialized right after this for `ImportNumericInitialData` where it retrieves the cache tags used by the numeric initial data importer. The non-specialized implementation does nothing. Do you think that's unclear and I should add a comment explaining what happens? I feel like this pattern of template specializations is common enough that we don't need to add a comment about that whenever we do that, don't you think?

geoffrey4444 commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -14,21 +14,111 @@
+                                        InitialData>::const_global_cache_tags;
+};
+
+template <typename DgElementArray, typename ImportInitialData>
```
Should the second parameter be named `ImportNoInitialData`?

nilsvu replied:
`ImportNoInitialData` is one possible class this template parameter could be set to, the other is `ImportNumericInitialData`. For the latter there's a template specialization below.

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -42,3 +42,14 @@ target_link_libraries(
   )
+
+add_generalized_harmonic_executable(
+  KerrSchildNumericInitialData
```
I'm not sure KerrSchildNumericInitialData is a good name, because it implies the numeric initial data is for KerrSchild. What if it's for perturbed KerrSchild, or for a binary? Maybe the name should just be EvolveGeneraliedHarmonicNumericInitialData? As long as the outer boundary is far enough away, the KerrSchild BC is still ok, at least for a little while (until GWs reach the outer boundary)

nilsvu replied:
Right, it's the GH executable with numeric initial data and KerrSchild boundary conditions. I'm happy for you and @nilsdeppe to choose a name for it :)

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/CMakeLists.txt:
```diff
@@ -42,3 +42,14 @@ target_link_libraries(
+target_link_libraries(
+  EvolveKerrSchildNumericInitialData
+  PRIVATE
+  GeneralRelativitySolutions
```
Since the analytic and numeric executables both do link against GeneralRelativitySolutions, why split it off?

nilsvu replied:
I figured since I'm changing the cmake file I might as well do that now. Once we have other boundary conditions the numeric initial data exec won't need to link the solutions anymore.

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -271,6 +275,9 @@ struct EvolutionMetavars {
     Initialization,
+    RegisterWithVolumeDataReader,
+    ImportInitialData,
+    InitializeDataDependentQuantities,
```
Dependent on what data? Maybe InitializeInitialDataDependentQuantities?

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -281,8 +288,10 @@ struct EvolutionMetavars {
       Initialization::Actions::NonconservativeSystem,
-      evolution::Initialization::Actions::SetVariables<
-          domain::Tags::Coordinates<volume_dim, Frame::Logical>>,
+      std::conditional_t<
```
So I can understand, please explain to me what this addition is doing and why it's necessary?

nilsvu replied:
This change makes sure we don't even try to set the initial data from the analytic solution, and therefore prevents initialization code below from using it (which would be a bug). If code before the `InitializeInitialDataDependentQuantities` phase tries to use initial data now we get an FPE (at least in Debug builds).

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -333,41 +342,60 @@ struct EvolutionMetavars {
       Initialization::Actions::AddComputeTags<
           tmpl::list<evolution::Tags::AnalyticCompute<
               volume_dim, analytic_solution_tag, analytic_solution_fields>>>,
-      GeneralizedHarmonic::Actions::InitializeDampedHarmonicRollonGauge<
```
How does this PR interact with #2237?

nilsvu replied:
I'm not sure, the PR you linked doesn't change any actions. If changes to these actions are coming up that interact with the initialization procedure perhaps we should talk briefly in a hangout

geoffrey4444 commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -333,41 +342,60 @@ struct EvolutionMetavars {
       Initialization::Actions::RemoveOptionsAndTerminatePhase>;
 
-  using component_list = tmpl::list<
+  using initialize_data_dependent_quantities_actions = tmpl::list<
```
initialize_initial_data_dependent_quantities_actions?

geoffrey4444 reviewed (changes requested):
My apologies for taking so long to review! I have questions and suggestions, but I think this PR is the way to go over #2005 for importing numeric initial data, so I closed #2005. Thanks for working on this and for rebasing!

I've reviewed the code, but I still want to test it with numeric initial data to make sure it works right in practice before final approval

geoffrey4444 commented:
Another question: is the numeric executable tested? If not, should it be?

nilsdeppe commented:
@geoffrey4444 do you have suggestions on how to test it? I'm not really a big fan of having a bunch of HDF5 files in the repo

geoffrey4444 commented:
> @geoffrey4444 do you have suggestions on how to test it? I'm not really a big fan of having a bunch of HDF5 files in the repo

Well maybe the test could write temporary data? Or maybe just an example input file somewhere would be sufficient?

geoffrey4444 commented:
@nilsleiffischer I tested the exec. It looks like it reads the initial data correctly but doesn’t seem to work with the Ah finder...I get a runtime error about non deductible template parameters / cannot call apply function.

Could you please verify you can get AH output when importing BH numerical data? Maybe just output KerrSchild from the analytic exec then read that in with the numeric exec

nilsvu commented:
@geoffrey4444 On this branch I can run `EvolveKerrSchild` with the input file from `tests` and then run `EvolveKerrSchildNumericInitialData` with these changes to the input file:

```diff
diff --git a/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml b/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
index 86b9e69b8..0914ef198 100644
--- a/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
+++ b/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
@@ -60,8 +60,8 @@ EventsAndTriggers:
   : - Completion
 
 Observers:
-  VolumeFileName: "GhKerrSchildVolumeData"
-  ReductionFileName: "GhKerrSchildReductionData"
+  VolumeFileName: "GhKerrSchildNumericIDVolumeData"
+  ReductionFileName: "GhKerrSchildNumericIDReductionData"
 
 ApparentHorizons:
   AhA:
@@ -72,3 +72,9 @@ ApparentHorizons:
     FastFlow:
       Flow: Fast
     Verbosity: Verbose
+
+Importers:
+  NumericInitialData:
+    FileName: GhKerrSchildVolumeData0.h5
+    Subgroup: element_data
+    ObservationValue: 0.
```

Both runs give the same output:

```
Charm++: standalone mode (not using charmrun)
Charm++> Running in Multicore mode:  1 threads
Converse/Charm++ Commit ID: v6.8.0-0-ga36028edb
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 unique compute nodes (12-way SMP).
Charm++> cpu topology info is gathered in 0.000 seconds.

Executing './bin/EvolveKerrSchildNumericInitialData' using 1 processors.
Date and time at startup: Tue May  5 18:35:59 2020

SpECTRE Build Information:
Version:                      0.0.0
Compiled on host:             nlf-mbp.local
Compiled in directory:        /Users/nlf/Work/spectre/build-Default-Debug
Source directory is:          /Users/nlf/Projects/spectre/develop
Compiled on git branch:       import_num_initial_data_in_gh
Compiled with git hash:       80cc98965f60565bb0b0571d8fe8c76041b7be11
Linked on:                    Tue May  5 18:34:02 2020

AhA: its=0: 2.0e-01<R<2e-01, |R|=0.2, |R_grid|=0.2, 2.2<r<2.2
AhA: its=1: 6.9e-02<R<1e-01, |R|=0.09, |R_grid|=0.09, 2.082<r<2.084
AhA: its=2: 2.1e-02<R<5e-02, |R|=0.04, |R_grid|=0.04, 2.034<r<2.036
AhA: its=3: -3.1e-04<R<3e-02, |R|=0.02, |R_grid|=0.02, 2.014<r<2.015
AhA: its=4: -9.5e-03<R<2e-02, |R|=0.008, |R_grid|=0.01, 2.004<r<2.006
AhA: its=5: -1.4e-02<R<1e-02, |R|=0.003, |R_grid|=0.008, 2<r<2.002
AhA: its=6: -1.5e-02<R<1e-02, |R|=0.002, |R_grid|=0.007, 1.998<r<2
AhA: its=7: -1.6e-02<R<9e-03, |R|=0.0007, |R_grid|=0.007, 1.997<r<1.999
AhA: its=8: -1.7e-02<R<9e-03, |R|=0.0003, |R_grid|=0.007, 1.997<r<1.999
AhA: its=9: -1.7e-02<R<8e-03, |R|=0.0001, |R_grid|=0.007, 1.997<r<1.999
AhA: its=10: -1.7e-02<R<8e-03, |R|=6e-05, |R_grid|=0.007, 1.997<r<1.999

Done!
CkWallTimer in seconds 1.788663
Date and time at completion: Tue May  5 18:36:01 2020

[Partition 0][Node 0] End of program
```

Do you encounter an issue with this? If you have an issue on another branch, could you point me to it? Then I can help you debug

geoffrey4444 commented:
> @geoffrey4444 On this branch I can run `EvolveKerrSchild` with the input file from `tests` and then run `EvolveKerrSchildNumericInitialData` with these changes to the input file:
> 
> 
> 
> ```diff
> 
> diff --git a/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml b/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
> 
> index 86b9e69b8..0914ef198 100644
> 
> --- a/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
> 
> +++ b/tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
> 
> @@ -60,8 +60,8 @@ EventsAndTriggers:
> 
>    : - Completion
> 
>  
> 
>  Observers:
> 
> -  VolumeFileName: "GhKerrSchildVolumeData"
> 
> -  ReductionFileName: "GhKerrSchildReductionData"
> 
> +  VolumeFileName: "GhKerrSchildNumericIDVolumeData"
> 
> +  ReductionFileName: "GhKerrSchildNumericIDReductionData"
> 
>  
> 
>  ApparentHorizons:
> 
>    AhA:
> 
> @@ -72,3 +72,9 @@ ApparentHorizons:
> 
>      FastFlow:
> 
>        Flow: Fast
> 
>      Verbosity: Verbose
> 
> +
> 
> +Importers:
> 
> +  NumericInitialData:
> 
> +    FileName: GhKerrSchildVolumeData0.h5
> 
> +    Subgroup: element_data
> 
> +    ObservationValue: 0.
> 
> ```
> 
> 
> 
> Both runs give the same output:
> 
> 
> 
> ```
> 
> Charm++: standalone mode (not using charmrun)
> 
> Charm++> Running in Multicore mode:  1 threads
> 
> Converse/Charm++ Commit ID: v6.8.0-0-ga36028edb
> 
> CharmLB> Load balancer assumes all CPUs are same.
> 
> Charm++> Running on 1 unique compute nodes (12-way SMP).
> 
> Charm++> cpu topology info is gathered in 0.000 seconds.
> 
> 
> 
> Executing './bin/EvolveKerrSchildNumericInitialData' using 1 processors.
> 
> Date and time at startup: Tue May  5 18:35:59 2020
> 
> 
> 
> SpECTRE Build Information:
> 
> Version:                      0.0.0
> 
> Compiled on host:             nlf-mbp.local
> 
> Compiled in directory:        /Users/nlf/Work/spectre/build-Default-Debug
> 
> Source directory is:          /Users/nlf/Projects/spectre/develop
> 
> Compiled on git branch:       import_num_initial_data_in_gh
> 
> Compiled with git hash:       80cc98965f60565bb0b0571d8fe8c76041b7be11
> 
> Linked on:                    Tue May  5 18:34:02 2020
> 
> 
> 
> AhA: its=0: 2.0e-01<R<2e-01, |R|=0.2, |R_grid|=0.2, 2.2<r<2.2
> 
> AhA: its=1: 6.9e-02<R<1e-01, |R|=0.09, |R_grid|=0.09, 2.082<r<2.084
> 
> AhA: its=2: 2.1e-02<R<5e-02, |R|=0.04, |R_grid|=0.04, 2.034<r<2.036
> 
> AhA: its=3: -3.1e-04<R<3e-02, |R|=0.02, |R_grid|=0.02, 2.014<r<2.015
> 
> AhA: its=4: -9.5e-03<R<2e-02, |R|=0.008, |R_grid|=0.01, 2.004<r<2.006
> 
> AhA: its=5: -1.4e-02<R<1e-02, |R|=0.003, |R_grid|=0.008, 2<r<2.002
> 
> AhA: its=6: -1.5e-02<R<1e-02, |R|=0.002, |R_grid|=0.007, 1.998<r<2
> 
> AhA: its=7: -1.6e-02<R<9e-03, |R|=0.0007, |R_grid|=0.007, 1.997<r<1.999
> 
> AhA: its=8: -1.7e-02<R<9e-03, |R|=0.0003, |R_grid|=0.007, 1.997<r<1.999
> 
> AhA: its=9: -1.7e-02<R<8e-03, |R|=0.0001, |R_grid|=0.007, 1.997<r<1.999
> 
> AhA: its=10: -1.7e-02<R<8e-03, |R|=6e-05, |R_grid|=0.007, 1.997<r<1.999
> 
> 
> 
> Done!
> 
> CkWallTimer in seconds 1.788663
> 
> Date and time at completion: Tue May  5 18:36:01 2020
> 
> 
> 
> [Partition 0][Node 0] End of program
> 
> ```
> 
> 
> 
> Do you encounter an issue with this? If you have an issue on another branch, could you point me to it? Then I can help you debug

I’ll check when I’m next able to, probably this afternoon. Have you tested running on multiple nodes?

nilsvu commented:
@geoffrey4444 @nilsdeppe about testing: I could add a test that does what I describe in the comment above - run with analytic initial data, then import the volume output as numeric initial data and run again. Not sure where to place such a test though, it could be in a directory next to `tests/Unit` and `tests/InputFiles` - any suggestions?

geoffrey4444 commented:
Here’s the branch I got the error with: https://github.com/sxs-collaboration/spectre/compare/develop...geoffrey4444:test1986?expand=1

geoffrey4444 commented:
@nilsleiffischer I did a clean build in release mode, and the AH finder appears to work.

geoffrey4444 commented:
Oops, spoke to soon: the Ah finder itself ran, but then I got this error, same as before: 

~~~~~
############ ERROR ############
Node: 3 Proc: 68
Line: 105 of /home/geoffrey/Codes/spectre/spectre/src/Parallel/SimpleActionVisitation.hpp
Function: void Parallel::Algorithm_detail::simple_action_visitor_helper(boost::variant<Variants...> &, const gsl::not_null<int *>, const gsl::not_null<bool *>, Args &&...) [Invokable = intrp::TargetPoints::ApparentHorizon<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, Frame::Inertial>, ParallelComponent = intrp::InterpolationTarget<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >, EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA>, ThisVariant = db::DataBox<brigand::list<Parallel::Tags::ConstGlobalCacheImpl<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, intrp::Tags::IndicesOfFilledInterpPoints<TimeStepId>, intrp::Tags::IndicesOfInvalidInterpPoints<TimeStepId>, intrp::Tags::TemporalIds<TimeStepId>, intrp::Tags::CompletedTemporalIds<TimeStepId>, intrp::Tags::InterpolatedVars<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, TimeStepId>, Tags::Variables<brigand::list<gr::Tags::SpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector>, gr::Tags::SpatialChristoffelSecondKind<3, Frame::Inertial, DataVector> > >, gr::Tags::SpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector>, gr::Tags::SpatialChristoffelSecondKind<3, Frame::Inertial, DataVector>, StrahlkorperTags::Strahlkorper<Frame::Inertial>, ah::Tags::FastFlow, Tags::Verbosity, Parallel::Tags::FromConstGlobalCache<Tags::AnalyticSolution<GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, Parallel::Tags::FromConstGlobalCache<Tags::NumericalFlux<GeneralizedHarmonic::UpwindFlux<3> > >, Parallel::Tags::FromConstGlobalCache<Tags::TimeStepper<TimeStepper> >, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHRollOnStartTime>, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHRollOnTimeWindow>, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHSpatialWeightDecayWidth<Frame::Inertial> >, Parallel::Tags::FromConstGlobalCache<Tags::EventsAndTriggers<brigand::list<Registration::Registrar<Events::ObserveErrorNorms, Tags::Time, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > >, dg::Events::Registrars::ObserveFields<3, Tags::Time, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial>, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::GaugeConstraint<3, Frame::Inertial> >, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::ThreeIndexConstraint<3, Frame::Inertial> >, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::FourIndexConstraint<3, Frame::Inertial> > >, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > >, Registration::Registrar<Events::ChangeSlabSize, brigand::list<StepChoosers::Registrars::Cfl<3, Frame::Inertial>, Registration::Registrar<StepChoosers::Constant>, Registration::Registrar<StepChoosers::Increase>, Registration::Registrar<StepChoosers::PreventRapidIncrease>, Registration::Registrar<StepChoosers::StepToTimes> > >, intrp::Events::Registrars::Interpolate<3, EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > > >, brigand::list<Registration::Registrar<Triggers::EveryNSlabs>, Registration::Registrar<Triggers::NearTimes>, Registration::Registrar<Triggers::PastTime>, Registration::Registrar<Triggers::SpecifiedSlabs>, Registration::Registrar<Triggers::SpecifiedTimes> > > >, Parallel::Tags::FromConstGlobalCache<observers::Tags::ReductionFileName>, Parallel::Tags::FromConstGlobalCache<observers::Tags::VolumeFileName>, Parallel::Tags::FromConstGlobalCache<intrp::Tags::ApparentHorizon<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, Frame::Inertial> >, Parallel::Tags::FromConstGlobalCache<domain::Tags::Domain<3> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::FileName<evolution::OptionTags::NumericInitialData> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::Subgroup<evolution::OptionTags::NumericInitialData> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::ObservationValue<evolution::OptionTags::NumericInitialData> >, StrahlkorperTags::ThetaPhi<Frame::Inertial>, StrahlkorperTags::Rhat<Frame::Inertial>, StrahlkorperTags::Jacobian<Frame::Inertial>, StrahlkorperTags::InvJacobian<Frame::Inertial>, StrahlkorperTags::InvHessian<Frame::Inertial>, StrahlkorperTags::Radius<Frame::Inertial>, StrahlkorperTags::CartesianCoords<Frame::Inertial>, StrahlkorperTags::DxRadius<Frame::Inertial>, StrahlkorperTags::D2xRadius<Frame::Inertial>, StrahlkorperTags::LaplacianRadius<Frame::Inertial>, StrahlkorperTags::NormalOneForm<Frame::Inertial>, StrahlkorperTags::Tangents<Frame::Inertial>, StrahlkorperGr::Tags::AreaElement<Frame::Inertial>, EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::Unity, StrahlkorperGr::Tags::SurfaceIntegral<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::Unity, Frame::Inertial> > >, Variants = <db::DataBox<brigand::list<> >, db::DataBox<brigand::list<Parallel::Tags::ConstGlobalCacheImpl<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, Parallel::Tags::FromConstGlobalCache<Tags::AnalyticSolution<GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, Parallel::Tags::FromConstGlobalCache<Tags::NumericalFlux<GeneralizedHarmonic::UpwindFlux<3> > >, Parallel::Tags::FromConstGlobalCache<Tags::TimeStepper<TimeStepper> >, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHRollOnStartTime>, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHRollOnTimeWindow>, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHSpatialWeightDecayWidth<Frame::Inertial> >, Parallel::Tags::FromConstGlobalCache<Tags::EventsAndTriggers<brigand::list<Registration::Registrar<Events::ObserveErrorNorms, Tags::Time, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > >, dg::Events::Registrars::ObserveFields<3, Tags::Time, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial>, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::GaugeConstraint<3, Frame::Inertial> >, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::ThreeIndexConstraint<3, Frame::Inertial> >, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::FourIndexConstraint<3, Frame::Inertial> > >, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > >, Registration::Registrar<Events::ChangeSlabSize, brigand::list<StepChoosers::Registrars::Cfl<3, Frame::Inertial>, Registration::Registrar<StepChoosers::Constant>, Registration::Registrar<StepChoosers::Increase>, Registration::Registrar<StepChoosers::PreventRapidIncrease>, Registration::Registrar<StepChoosers::StepToTimes> > >, intrp::Events::Registrars::Interpolate<3, EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > > >, brigand::list<Registration::Registrar<Triggers::EveryNSlabs>, Registration::Registrar<Triggers::NearTimes>, Registration::Registrar<Triggers::PastTime>, Registration::Registrar<Triggers::SpecifiedSlabs>, Registration::Registrar<Triggers::SpecifiedTimes> > > >, Parallel::Tags::FromConstGlobalCache<observers::Tags::ReductionFileName>, Parallel::Tags::FromConstGlobalCache<observers::Tags::VolumeFileName>, Parallel::Tags::FromConstGlobalCache<intrp::Tags::ApparentHorizon<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, Frame::Inertial> >, Parallel::Tags::FromConstGlobalCache<domain::Tags::Domain<3> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::FileName<evolution::OptionTags::NumericInitialData> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::Subgroup<evolution::OptionTags::NumericInitialData> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::ObservationValue<evolution::OptionTags::NumericInitialData> > > >, db::DataBox<brigand::list<Parallel::Tags::ConstGlobalCacheImpl<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, intrp::Tags::IndicesOfFilledInterpPoints<TimeStepId>, intrp::Tags::IndicesOfInvalidInterpPoints<TimeStepId>, intrp::Tags::TemporalIds<TimeStepId>, intrp::Tags::CompletedTemporalIds<TimeStepId>, intrp::Tags::InterpolatedVars<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, TimeStepId>, Tags::Variables<brigand::list<gr::Tags::SpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector>, gr::Tags::SpatialChristoffelSecondKind<3, Frame::Inertial, DataVector> > >, gr::Tags::SpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>, gr::Tags::ExtrinsicCurvature<3, Frame::Inertial, DataVector>, gr::Tags::SpatialChristoffelSecondKind<3, Frame::Inertial, DataVector>, StrahlkorperTags::Strahlkorper<Frame::Inertial>, ah::Tags::FastFlow, Tags::Verbosity, Parallel::Tags::FromConstGlobalCache<Tags::AnalyticSolution<GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, Parallel::Tags::FromConstGlobalCache<Tags::NumericalFlux<GeneralizedHarmonic::UpwindFlux<3> > >, Parallel::Tags::FromConstGlobalCache<Tags::TimeStepper<TimeStepper> >, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHRollOnStartTime>, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHRollOnTimeWindow>, Parallel::Tags::FromConstGlobalCache<GeneralizedHarmonic::Tags::GaugeHSpatialWeightDecayWidth<Frame::Inertial> >, Parallel::Tags::FromConstGlobalCache<Tags::EventsAndTriggers<brigand::list<Registration::Registrar<Events::ObserveErrorNorms, Tags::Time, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > >, dg::Events::Registrars::ObserveFields<3, Tags::Time, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial>, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::GaugeConstraint<3, Frame::Inertial> >, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::ThreeIndexConstraint<3, Frame::Inertial> >, Tags::PointwiseL2Norm<GeneralizedHarmonic::Tags::FourIndexConstraint<3, Frame::Inertial> > >, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > >, Registration::Registrar<Events::ChangeSlabSize, brigand::list<StepChoosers::Registrars::Cfl<3, Frame::Inertial>, Registration::Registrar<StepChoosers::Constant>, Registration::Registrar<StepChoosers::Increase>, Registration::Registrar<StepChoosers::PreventRapidIncrease>, Registration::Registrar<StepChoosers::StepToTimes> > >, intrp::Events::Registrars::Interpolate<3, EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, brigand::list<gr::Tags::SpacetimeMetric<3, Frame::Inertial, DataVector>, GeneralizedHarmonic::Tags::Pi<3, Frame::Inertial>, GeneralizedHarmonic::Tags::Phi<3, Frame::Inertial> > > >, brigand::list<Registration::Registrar<Triggers::EveryNSlabs>, Registration::Registrar<Triggers::NearTimes>, Registration::Registrar<Triggers::PastTime>, Registration::Registrar<Triggers::SpecifiedSlabs>, Registration::Registrar<Triggers::SpecifiedTimes> > > >, Parallel::Tags::FromConstGlobalCache<observers::Tags::ReductionFileName>, Parallel::Tags::FromConstGlobalCache<observers::Tags::VolumeFileName>, Parallel::Tags::FromConstGlobalCache<intrp::Tags::ApparentHorizon<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, Frame::Inertial> >, Parallel::Tags::FromConstGlobalCache<domain::Tags::Domain<3> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::FileName<evolution::OptionTags::NumericInitialData> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::Subgroup<evolution::OptionTags::NumericInitialData> >, Parallel::Tags::FromConstGlobalCache<importers::Tags::ObservationValue<evolution::OptionTags::NumericInitialData> >, StrahlkorperTags::ThetaPhi<Frame::Inertial>, StrahlkorperTags::Rhat<Frame::Inertial>, StrahlkorperTags::Jacobian<Frame::Inertial>, StrahlkorperTags::InvJacobian<Frame::Inertial>, StrahlkorperTags::InvHessian<Frame::Inertial>, StrahlkorperTags::Radius<Frame::Inertial>, StrahlkorperTags::CartesianCoords<Frame::Inertial>, StrahlkorperTags::DxRadius<Frame::Inertial>, StrahlkorperTags::D2xRadius<Frame::Inertial>, StrahlkorperTags::LaplacianRadius<Frame::Inertial>, StrahlkorperTags::NormalOneForm<Frame::Inertial>, StrahlkorperTags::Tangents<Frame::Inertial>, StrahlkorperGr::Tags::AreaElement<Frame::Inertial>, EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::Unity, StrahlkorperGr::Tags::SurfaceIntegral<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::Unity, Frame::Inertial> > >>, Args = <Parallel::ConstGlobalCache<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > > &, const cpp17::void_type &, TimeStepId>]

Cannot call apply function of 'intrp::TargetPoints::ApparentHorizon<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3ul> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> >::AhA, Frame::Inertial>' with DataBox type 'N2db7DataBoxIN7brigand4listIJN8Parallel4Tags20ConstGlobalCacheImplI17EvolutionMetavarsIN9evolution18NumericInitialDataIN19GeneralizedHarmonic6SystemILm3EEEEENS9_9Solutions9WrappedGrIN2gr9Solutions10KerrSchildEEEEEEN5intrp4Tags27IndicesOfFilledInterpPointsI10TimeStepIdEENSM_28IndicesOfInvalidInterpPointsISO_EENSM_11TemporalIdsISO_EENSM_20CompletedTemporalIdsISO_EENSM_16InterpolatedVarsINSJ_3AhAESO_EEN4Tags9VariablesINS2_IJNSF_4Tags13SpatialMetricILm3EN5Frame8InertialE10DataVectorEENS11_20InverseSpatialMetricILm3ES14_S15_EENS11_18ExtrinsicCurvatureILm3ES14_S15_EENS11_28SpatialChristoffelSecondKindILm3ES14_S15_EEEEEEES16_S18_S1A_S1C_N16StrahlkorperTags12StrahlkorperIS14_EEN2ah4Tags8FastFlowENSZ_9VerbosityENS4_20FromConstGlobalCacheINSZ_16AnalyticSolutionISI_EEEENS1M_INSZ_13NumericalFluxINS9_10UpwindFluxILm3EEEEEEENS1M_INSZ_11TimeStepperI11TimeStepperEEEENS1M_INS9_4Tags21GaugeHRollOnStartTimeEEENS1M_INS1Z_22GaugeHRollOnTimeWindowEEENS1M_INS1Z_29GaugeHSpatialWeightDecayWidthIS14_EEEENS1M_INSZ_17EventsAndTriggersINS2_IJN12Registration9RegistrarIN2dg6Events17ObserveErrorNormsEJNSZ_4TimeENS2_IJNS11_15SpacetimeMetricILm3ES14_S15_EENS1Z_2PiILm3ES14_EENS1Z_3PhiILm3ES14_EEEEEEEENS2B_10Registrars13ObserveFieldsILm3ES2D_NS2_IJS2F_S2H_S2J_NSZ_15PointwiseL2NormINS1Z_15GaugeConstraintILm3ES14_EEEENS2O_INS1Z_20ThreeIndexConstraintILm3ES14_EEEENS2O_INS1Z_19FourIndexConstraintILm3ES14_EEEEEEES2K_EENS29_IN6Events14ChangeSlabSizeEJNS2_IJN12StepChoosers10Registrars3CflILm3ES14_EENS29_INS32_8ConstantEJEEENS29_INS32_8IncreaseEJEEENS29_INS32_20PreventRapidIncreaseEJEEENS29_INS32_11StepToTimesEJEEEEEEEEENSL_6Events10Registrars11InterpolateILm3ESX_S2K_EEEEENS2_IJNS29_IN8Triggers11EveryNSlabsEJEEENS29_INS3L_9NearTimesEJEEENS29_INS3L_8PastTimeEJEEENS29_INS3L_14SpecifiedSlabsEJEEENS29_INS3L_14SpecifiedTimesEJEEEEEEEEEENS1M_IN9observers4Tags17ReductionFileNameEEENS1M_INS40_14VolumeFileNameEEENS1M_INSM_15ApparentHorizonISX_S14_EEEENS1M_IN6domain4Tags6DomainILm3EEEEENS1M_IN9importers4Tags8FileNameINS7_10OptionTags18NumericInitialDataEEEEENS1M_INS4E_8SubgroupIS4H_EEEENS1M_INS4E_16ObservationValueIS4H_EEEENS1F_8ThetaPhiIS14_EENS1F_4RhatIS14_EENS1F_8JacobianIS14_EENS1F_11InvJacobianIS14_EENS1F_10InvHessianIS14_EENS1F_6RadiusIS14_EENS1F_15CartesianCoordsIS14_EENS1F_8DxRadiusIS14_EENS1F_9D2xRadiusIS14_EENS1F_15LaplacianRadiusIS14_EENS1F_13NormalOneFormIS14_EENS1F_8TangentsIS14_EEN14StrahlkorperGr4Tags11AreaElementIS14_EENSJ_5UnityENS5F_15SurfaceIntegralIS5I_S14_EEEEEEE' and arguments 'Parallel::ConstGlobalCache<EvolutionMetavars<evolution::NumericInitialData<GeneralizedHarmonic::System<3ul> >, GeneralizedHarmonic::Solutions::WrappedGr<gr::Solutions::KerrSchild> > >, cpp17::void_type, TimeStepId'.
If the argument types to the apply function match, then it is possible that the apply function has non-deducible template parameters. This could occur from removing an apply function argument and forgetting to remove its associated template parameters.

############ ERROR ############
~~~~~

geoffrey4444 commented:
> @nilsleiffischer I did a clean build in release mode, and the AH finder appears to work.

The error appears to be related to the interpolator, not this PR. @nilsleiffischer and I discovered it also triggers for analytic initial data if you run the AH finder more than once. So I don't think there's anything to worry about in terms of this PR

geoffrey4444 commented:
@nilsleiffischer Aside from the comments I made in my review. I’m now happy with this PR. It successfully loads numeric data.

geoffrey4444 commented:
Thanks for considering my comments @nilsleiffischer! I’m happy for you to squash

geoffrey4444 commented:
Merge conflicts...please rebase?

geoffrey4444 commented:
Might a core developer review this, whenever one of you gets a chance to take a look?
@nilsdeppe @kidder @wthrowe

geoffrey4444 commented:
@nilsleiffischer needs a rebase

wthrowe commented:
This looks fine to me.  Just needs the rebase.

geoffrey4444 reviewed (dismissed):
I presume the clang-tidy failure is one we’ll ignore?

nilsdeppe commented:
We've only been ignorning clang-tidy failures when they're being addressed in another PR. Since this one is not, it should be fixed here

nilsvu committed:
Fix a clang-tidy warning

geoffrey4444 commented:
Please squash and ensure all tests pass

nilsdeppe commented on src/Evolution/Protocols.hpp:
```diff
@@ -0,0 +1,33 @@
+// Distributed under the MIT License.
```
Until the discussion with whether or not to continue to have protocols in the (evolution) code is settled, please stop introducing them.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -37,7 +127,10 @@ struct DgElementArray {
-  using const_global_cache_tags = tmpl::list<domain::Tags::Domain<volume_dim>>;
+  using const_global_cache_tags = tmpl::flatten<tmpl::list<
+      domain::Tags::Domain<volume_dim>,
+      tmpl::type_from<DgElementArray_detail::import_numeric_data_cache_tags<
```
@nilsleiffischer can you please remove the `type_from` here?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/DgElementArray.hpp:
```diff
@@ -14,21 +14,111 @@
+    ImportNumericInitialData<PhaseType, ImportPhase, InitialData>> {
+  template <typename Metavariables>
+  static void apply(
+      const typename Metavariables::Phase next_phase,
```
Since you have `Metavariables` in this function, is it not possible to instead check from the metavariables if numerical data is being imported? Then you can use `if constexpr` instead of the specialization.

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -284,8 +291,10 @@ struct EvolutionMetavars {
       Initialization::Actions::NonconservativeSystem,
-      evolution::Initialization::Actions::SetVariables<
-          domain::Tags::Coordinates<volume_dim, Frame::Logical>>,
+      std::conditional_t<
```
Can you switch this to `tmpl::conditional_t`

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -326,41 +335,60 @@ struct EvolutionMetavars {
       observers::ObserverWriter<EvolutionMetavars>,
       intrp::Interpolator<EvolutionMetavars>,
       intrp::InterpolationTarget<EvolutionMetavars, AhA>,
+      std::conditional_t<evolution::is_numeric_initial_data_v<initial_data>,
```
`tmpl::conditional_t`

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -326,41 +335,60 @@ struct EvolutionMetavars {
               Parallel::PhaseActions<Phase, Phase::Initialization,
                                      initialization_actions>,
-
+              std::conditional_t<
```
`tmpl::conditional_t`

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -326,41 +335,60 @@ struct EvolutionMetavars {
+      GeneralizedHarmonic::Actions::InitializeConstraints<volume_dim>,
+      Parallel::Actions::TerminatePhase>;
+
+  using component_list = tmpl::flatten<tmpl::list<
```
do we need the `flatten` still here?

nilsvu replied:
The `component_list` is not currently flattened in the backend, it's being retrieved directly from the `Metavariables` in multiple places

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -326,41 +335,60 @@ struct EvolutionMetavars {
       DgElementArray<
           EvolutionMetavars,
-          tmpl::list<
+          tmpl::flatten<tmpl::list<
```
Is this a place we didn't add `flatten` in the backend?

nilsvu replied:
The PDAL is also not currently flattened in the backend. Only the list of actions in each phase is flattened.

nilsdeppe commented on src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonic.hpp:
```diff
@@ -326,41 +335,60 @@ struct EvolutionMetavars {
-                             step_actions, Actions::AdvanceTime>>>>>;
+                             Actions::ChangeSlabSize, step_actions,
+                             Actions::AdvanceTime>>>>,
+          std::conditional_t<evolution::is_numeric_initial_data_v<initial_data>,
```
`tmpl::conditional_t`

nilsvu commented:
@nilsdeppe Adressed your comments in #2337
